### PR TITLE
Expose semantic tokens functionality to Razor EA

### PR DESCRIPTION
**Pull request by davidwengier**
As part of https://github.com/dotnet/razor/issues/9519 we need to be able to call into the guts of the Roslyn LSP handlers via C#, so this PR exposes semantic tokens as the first cab off the rank. In the meeting about this we said we were okay with using LSP types in these methods, but for semantic tokens specifically it seemed easy enough to avoid that, as the only relevant type is `Range` and it already has a good analog in `LinePositionSpan`.

There is more work to come with this I'm sure, but this will unblock things in Razor a little.

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs [hunk 1]
```diff
@@ -36,60 +35,82 @@ internal class SemanticTokensHelpers
...
+
+            // The above call to get semantic tokens may be inaccurate (because we use frozen partial semantics).  Kick
+            // off a request to ensure that the OOP side gets a fully up to compilation for this project.  Once it does
+            // we can optionally choose to notify our caller to do a refresh if we computed a compilation for a new
+            // solution snapshot.
+            await semanticTokensRefreshQueue.TryEnqueueRefreshComputationAsync(project, cancellationToken).ConfigureAwait(false);
+            return tokensData;
+        }
+
+        public static async Task<int[]> HandleRequestHelperAsync( Document document, ImmutableArray<LinePositionSpan> spans, bool supportsVisualStudioExtensions, ClassificationOptions options, CancellationToken cancellationToken)
```
Comment (davidwengier) [inline src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:58]:
```suggestion
        public static async Task<int[]> HandleRequestHelperAsync(Document document, ImmutableArray<LinePositionSpan> spans, bool supportsVisualStudioExtensions, ClassificationOptions options, CancellationToken cancellationToken)
```

**Comment by davidwengier** [comment]
ping @CyrusNajmabadi @dibarbet @dotnet/roslyn-ide, reviews would be much appreciated so this can be consumed from Razor.

File: src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs [hunk 2]
```diff
@@ -118,13 +120,14 @@ public class Template
                 markup, mutatingLspWorkspace, GetCapabilities(isVS));
 
             var document = testLspServer.GetCurrentSolution().Projects.First().Documents.First();
-            var ranges = new LSP.Range[2] {
-                new LSP.Range { Start = new Position(12, 0), End = new Position(13, 0) },
-                new LSP.Range { Start = new Position(29, 0), End = new Position(30, 0) }
-            };
+            ImmutableArray<LinePositionSpan> spans = [
```
Comment (dibarbet) [inline src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs:123]: Hmm, feels like these tests should be calling the actual range handler..., but thats not a new problem in this PR

File: src/Tools/ExternalAccess/Razor/Cohost/Handlers/SemanticTokensRange.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                        spans,
+                        supportsVisualStudioExtensions,
+                        ClassificationOptions.Default,
+                        cancellationToken);
+
+            // The above call to get semantic tokens may be inaccurate (because we use frozen partial semantics).  Kick
+            // off a request to ensure that the OOP side gets a fully up to compilation for this project.  Once it does
+            // we can optionally choose to notify our caller to do a refresh if we computed a compilation for a new
+            // solution snapshot.
+            // TODO: await semanticTokensRefreshQueue.TryEnqueueRefreshComputationAsync(project, cancellationToken).ConfigureAwait(false);
```
Comment (dibarbet) [inline src/Tools/ExternalAccess/Razor/Cohost/Handlers/SemanticTokensRange.cs:33]: why was this commented out?

**Review by dibarbet — APPROVED** [review]

File: src/Tools/ExternalAccess/Razor/Cohost/Handlers/SemanticTokensRange.cs (hunk 3 above)
Comment (davidwengier) [reply src/Tools/ExternalAccess/Razor/Cohost/Handlers/SemanticTokensRange.cs:33]: This is on the long list of follow-ups.

This method is called in the OOP process, so the queue has no access to send a refresh notification back to the client from here. We'll have to expose the queue functionality to the actual endpoint, and handle it there. Razor also has code to intercept refreshes, and map them from the generated C# back to the Razor file, and we need to work out a way to do that (if its still relevant? maybe not for semantic tokens)

**Comment by davidwengier** [comment]
Thanks David! Merging so I can get an updated package flowing. Happy to follow up if anyone has any further comments (that hopefully aren't binary breaking 😛)

**Pull request merged**